### DISCUSSION TITLE
Issues loading python 3.11 .dll/.so when trying to run a python script from a .NET 6 app hosted in a docker container

### DISCUSSION THREAD

**Discussion by Andrew-Rominger** [pythonnet/pythonnet#2046] [General]
### Environment

-   Pythonnet version: 3.0.1
-   Python version: 3.11
-   Operating System: Windows/Linux Docker Container
-   .NET Runtime: .NET 6.0

### Details

-   Describe what you were trying to get done.

    I am trying to get a .net 6.0 app to run python scripts when running in a docker container. 
    My code works fine running on Windows 10 or Ubuntu on WSL, but once I build it into a container I get an exception when I call PythonEngine.Initialize()

    ![image](https://user-images.githubusercontent.com/16858466/206916950-b7bf4fc2-1095-4701-a9bc-5626c6ce7dd4.png)

    I have triple checked that the python files get put in the right spot in my docker image

    ![image](https://user-images.githubusercontent.com/16858466/206917404-d62ec003-9ffb-45e2-8958-b8ee832b9b73.png)

-   What commands did you run to trigger this issue? If you can provide a
    [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
    this will help us understand the issue.

    I have set Runtime.PythonDLL to point at the .dll before calling Initialize
   
    ![image](https://user-images.githubusercontent.com/16858466/206917786-726640b4-890c-4552-b5b8-701474bf45f8.png)

    This works when I run the program directly on my Windows 10 pc, and through WSL after changing the path to point at the .so file in /usr/lib/python3.11/config-3.11-x86_64-linux-gnu/
    I have tried both linux and windows containers and I get the same exception saying it cant load the dynamic library. I am basically brand new to using Docker so I am assuming I have made some stupid mistake, here is the dockerfile I used for Windows, basically combining some stuff from the .NET 6 example, and python example files

    ```
    FROM mcr.microsoft.com/windows:10.0.17763.3650-amd64 as win
    
    RUN powershell.exe -Command $ErrorActionPreference = 'Stop'; \
        [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; \
        wget https://www.python.org/ftp/python/3.11.1/python-3.11.1-amd64.exe -OutFile c:\python-3.11.1-amd64.exe ; \
        Start-Process c:\python-3.11.1-amd64.exe -ArgumentList '/quiet PrependPath=1 TargetDir=C:\Python311' -Wait ; \
        Remove-Item c:\python-3.11.1-amd64.exe -Force
    
    FROM mcr.microsoft.com/dotnet/sdk:6.0 as build-env
    WORKDIR /src
    COPY src/*.csproj .
    RUN dotnet restore
    COPY src .
    RUN dotnet publish -c Release -o /publish
    
    FROM mcr.microsoft.com/dotnet/runtime:6.0 as runtime
    WORKDIR /app
    COPY --from=build-env /publish .
    COPY --from=win C:/Python311 C:/Python311
    ENTRYPOINT ["dotnet", "TestPythonNetDocker.dll"]
    ```

    and for linux. I am not used to working with linux and getting python installed and moved to the image felt harder than it needed to   be which means I am probably missing something.

 
    ```
    # syntax=docker/dockerfile:1 
    FROM ubuntu:jammy as ubu
    RUN apt-get update
    RUN apt-get install -y python3.11 libpython3.11
    
    FROM mcr.microsoft.com/dotnet/sdk:6.0 as build-env
    WORKDIR /src
    COPY src/*.csproj .
    RUN dotnet restore
    COPY src .
    RUN dotnet publish -c Release -o /publish
    
    FROM mcr.microsoft.com/dotnet/runtime:6.0 as runtime
    WORKDIR /app
    COPY --from=build-env /publish .
    COPY --from=ubu /usr/lib/python3.11/ /usr/lib/python3.11/
    RUN apt-get update
    ENTRYPOINT ["dotnet", "TestPythonNetDocker.dll"]
    ```

    I looked around and found some other issues about getting a python program to import .NET stuff in docker containers but didn't see anything about the otherway around.

**Comment by Andrew-Rominger** [comment]
I got it working by copying the entire python3.11 folder from /usr/lib to my source dir, added it to the project so it would be copied to the output dir, and referenced it directly like `python3.11/config-3.11-x86_64-linux-gnu/libpython3.11.so `
So it is an issue with my docker setup but that is not python.net's problem

**Reply by cricketthomas** [reply]
Thanks for your follow up! I am doing exactly what you are doing, python in a ubuntu docker container:
```
using Python.Runtime;

string pythonDll = @"/usr/lib/python3.8/config-3.8-x86_64-linux-gnu/libpython3.8.so";
Environment.SetEnvironmentVariable("PYTHONNET_PYDLL", pythonDll);
PythonEngine.Initialize();

using var python = Py.GIL();
using var scope = Py.CreateScope();
string code = System.IO.File.ReadAllText(@"/mnt/c/repos/mycode/test/python/example.py");
scope.Exec(code);

object results = scope.Get<object>("results");
object json = scope.Get<object>("json");

Console.WriteLine(json);
```
